aws/resource_aws_ec2_client_vpn_endpoint.go

```diff
@@ -0,0 +1,347 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsEc2ClientVpnEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsEc2ClientVpnEndpointCreate,
+		Read:   resourceAwsEc2ClientVpnEndpointRead,
+		Delete: resourceAwsEc2ClientVpnEndpointDelete,
+		Update: resourceAwsEc2ClientVpnEndpointUpdate,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"client_cidr_block": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"dns_servers": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"server_certificate_arn": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"transport_protocol": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Default:      "udp",
+				ValidateFunc: validation.StringInSlice([]string{"udp", "tcp"}, false),
+			},
+			"authentication_options": {
+				Type:     schema.TypeSet,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							ValidateFunc: validation.StringInSlice([]string{"certificate-authentication", "directory-service-authentication"}, false),
+						},
+						"active_directory_id": {
```

Comment:
This attribute does not currently have an associated acceptance test.

Reply:
I added a new acceptance test to check for using the AD based authentication option.